lib/util.js

```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
 // process.versions needs a custom function as some values are lazy-evaluated.
 process.versions[exports.inspect.custom] =
   (depth) => exports.format(JSON.parse(JSON.stringify(process.versions)));
+
+const kCustomCallbackifyiedSymbol = Symbol('util.callbackify.custom');
+const kCustomCallbackifyiedToSplat = Symbol('util.callbackify.toSplat');
+const kCustomCallbackifyiedProject = Symbol('util.callbackify.project');
+
+function callbackify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to callbackify() ' +
+      'must be a function');
+  }
+
+  if (orig[kCustomCallbackifyiedSymbol]) {
+    const fn = orig[kCustomCallbackifyiedSymbol];
+    if (typeof fn !== 'function') {
+      throw new TypeError('The [util.callbackify.custom] property must be ' +
+        'a function');
+    }
+    Object.defineProperty(fn, kCustomCallbackifyiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  // we have two way to convert the returned value into multiple arguments
+  // either splat it or project some of it's keys
+  const toSplat = orig[kCustomCallbackifyiedToSplat];
+  const project = orig[kCustomCallbackifyiedProject];
+
+  const fn = function(...args) {
+    const cb = args.pop();
+    if (typeof cb !== 'function') {
+      throw new TypeError('The first argument to the callbackified function ' +
+        'must be a callback style function');
+    }
+    orig.call(this, ...args).then((ret) => {
+      if (toSplat) {
+        cb(null, ...ret);
```

Comment:
So it'll go out of the Promise timeline?
I'd rather explicitly explode a la https://github.com/nodejs/node/pull/12712/files#diff-3deb3f32958bb937ae05c6f3e4abbdf5R1121